authenticate-users/test/FunctionsTest.php

```diff
@@ -0,0 +1,48 @@
+<?php
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Test for application controllers
+ */
+class ControllersTest extends TestCase
+{
+    private static $collection;
+
+    public static function setUpBeforeClass()
+    {
+        $_SERVER['REQUEST_URI'] = '/';
+        ob_start();
+        // Use output buffers to prevent print statements from showing up in test output.
+        require_once __DIR__ . '/../index.php';
+        ob_clean();
+    }
+
+    public function testInvalidJwt()
+    {
+        validate_assertion('fake_jwt', 'fake_expected_audience');
+        $this->expectOutputRegex('/Failed to validate assertion: Cannot decode compact serialisation/');
+    }
+
+    public function testCerts()
+    {
+        $certs = certs();
```

Comment:
Are we sure we want this test executing network requests?

Reply:
Why not? Are you concerned about potential failures due to network flakiness?

Reply:
Yes, that's part of it. Mocking would also facilitate offline development and would make it easier to test the get_metadata calls when you're not on app/compute engine.

Reply:
I think it'd be difficult to inject mock objects into the sample code without complicating the sample. For instance, we could do something like this:

```php
function certs(GuzzleHttp\ClientInterface $client = null) : string
{
    $client = $client ?: new GuzzleHttp\Client();
    $response = $client->get(
        'https://www.gstatic.com/iap/verify/public_key-jwk'
    );
    return $response->getBody();
}
```

But here I think the disadvantage of sample complexity outweighs the advantage of mocking the function. I'm open to it, I'm just not sure how we would go about it.